src/App.tsx

```diff
@@ -1,24 +1,216 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
-import { UserWarning } from './UserWarning';
+import React, { useCallback, useEffect, useState } from 'react';
+import { UserWarning } from './components/UserWarning';
+import {
+  getTodosFromServer,
+  addTodoOnServer,
+  deleteTodoOnServer,
+  updateTodoOnServer,
+} from './api/todos';
 
-const USER_ID = 0;
+import { TodoHeader } from './components/TodoHeader';
+import { TodoList } from './components/TodoList';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorNotification } from './components/ErrorNotification';
+
+import { Todo } from './types/Todo';
+import { TodoMode } from './types/TodoMode';
+import { TodoCompletionType } from './types/TodoCompletionType';
+import { ErrorType } from './types/ErrorType';
+import { TodoRich } from './types/TodoRich';
+import { TodoRichEditable } from './types/TodoRichEditable';
+
+import { filterTodos, getActiveTodosCount } from './common/helpers';
+
+const USER_ID = 6955;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<TodoRich[]>([]);
+  const [error, setError] = useState<ErrorType>(ErrorType.None);
+  const [filterType, setFilterType] = useState(TodoCompletionType.All);
+
+  const [tempTodo, setTempTodo] = useState<TodoRich | null>(null);
+
+  const addTodoLocal = (newTodo: Todo): void => {
+    setTodos((prevTodos) => (
+      [
+        ...prevTodos,
+        {
+          ...newTodo,
+          mode: TodoMode.None,
+        },
+      ]
+    ));
+  };
+
+  const deleteTodoLocal = (todoId: number): void => {
+    setTodos(prevTodos => (
+      prevTodos.filter(todo => todo.id !== todoId)
+    ));
+  };
+
+  const updateTodoLocal = (
+    todoId: number,
+    updatedData: TodoRichEditable,
+  ) => (
+    setTodos(prevTodos => (
+      prevTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return {
+          ...todo,
+          ...updatedData,
+        };
+      })
+    ))
+  );
+
+  const handleTodoAdd = async (todoTitle: string) => {
+    if (todoTitle === '') {
+      setError(ErrorType.EmptyTitle);
+
+      return;
+    }
+
+    const newTodo: TodoRich = {
+      id: 0,
+      title: todoTitle,
+      completed: false,
+      userId: USER_ID,
+      mode: TodoMode.Loading,
+    };
+
+    setTempTodo(newTodo);
+
+    try {
+      const newTodoFromServer = await addTodoOnServer(USER_ID, newTodo);
+
+      addTodoLocal(newTodoFromServer);
+    } catch {
+      setError(ErrorType.AddTodo);
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const handleTodoDelete = async (todoId: number) => {
+    updateTodoLocal(todoId, { mode: TodoMode.Loading });
+
+    try {
+      await deleteTodoOnServer(USER_ID, todoId);
+      deleteTodoLocal(todoId);
+    } catch {
+      setError(ErrorType.DeleteTodo);
+      updateTodoLocal(todoId, { mode: TodoMode.None });
```

Comment:
But why should we update todo if we removed it?

Reply:
This statement is in `catch` block - if we didn't remove `todo` in case of an error, we will leave the `todo` item in the list and just change its mode to `None`